### PR TITLE
Improve hooks ordering behavior

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90105
* #91667
* #91475
* __->__ #85849

Addresses: https://github.com/pytorch/pytorch/issues/35802

Design doc: https://docs.google.com/document/d/19xSib7FFknRQ5f3ptGFUmiOt3BrgXSUlTQH2xMcZJYg/edit#

#### Implementation
- We have now have 3 fields: pre_hooks, retains_grad_hooks, and tensor_pre_hooks so that we can more precisely define their ordering and when they are executed.
- Since retains grad uses an entirely new field, we cannot reuse the old retains grad, logic. We refactor retains grad to call directly into the variable.cpp logic. Other logic in variable.cpp that handle cpp hooks must also be updated. 

#### Hooks ordering and execution:
- Defines pre-hooks registered on tensor to run before pre-hooks registered on grad_fn
- Updates pre-hooks registered on tensor to always run, even if they are the inputs= to .grad()
- Post hooks (and pre hooks) can now observe the modifications to gradient by the tensor pre hook

#### Retains grad hooks
- retains grad hooks always execute last, even if there are other tensor pre-hooks registered

#### Unchanged:
- pre_hooks registered to grad_fn aren't expected to execute if they are the inputs= to .grad()

Follow ups:
- simplify retains_grad field to not be a vector, since it always holds a single hook
- potentially merge capture hooks with tensor pre hooks, this would involve some additional refactoring since 
- python hooks registered to tensor behavior on in-place is still wrong

cc @ezyang @gchanan

### CODE REVIEW

**Comment by soulitzer** [comment]
Update on how pre-hooks are ordered:

In this PR I decide to order pre-hooks before tensor pre-hooks for simplicity:
- The main reason is that retains_grad is a tensor pre-hook used in implementation of `backward(inputs=)` and we'd like its behavior to be consistent with that of `.grad(inputs=)`, i.e. the captured grad should reflect changes by all pre hooks
- I don't think this ordering matters too much in user code since if a user really wanted to read all the gradients at once, and still modify gradients on a per-tensor basis, the workaround is just to do everything on the grad_fn hook
- I am open to changing my mind if it can be shown that READS tend to occur on hooks to grad_fn and WRITES tend to occur on hooks to tensor (we want writes to happen before reads). However, it is tricky to proceed this way:
   - either we'd have to be OK with the inconsistency between .backward and .grad
   - or, we'd need to special case retains_grad to be called with the rest of the pre-hooks registered to grad_fn (increases complexity significantly)

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -883,8 +893,8 @@ static variable_list call_function(
   CheckpointValidGuard cpvguard(graph_task);
   auto& fn = *func;
   auto inputs =
-      call_pre_hooks(fn, InputBuffer::variables(std::move(inputBuffer)));
-
+      call_tensor_pre_hooks(fn, InputBuffer::variables(std::move(inputBuffer)));
+  inputs = call_pre_hooks(*func, inputs);
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:897]: nit: *func -> fn
also std::move(inputs)

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 2]
```diff
@@ -42,13 +42,10 @@ struct TORCH_API AccumulateGrad : public Node {
 
   variable_list apply(variable_list&& grads) override;
 
-  static at::Tensor callHooks(const Variable& variable, at::Tensor new_grad) {
-    for (auto& hook : impl::hooks(variable)) {
-      new_grad = (*hook)({new_grad})[0];
-    }
-    return new_grad;
+  std::vector<std::unique_ptr<FunctionPreHook>>& tensor_pre_hooks() noexcept
```
Comment (albanD) [inline torch/csrc/autograd/functions/accumulate_grad.h:45]: nit could you leave a commend here that says:
Since the AccumulateGrad Node is only a weak ref from the Tensor, it can be destroyed even though the Tensor is still alive (contrary to all other Nodes). So we must lazily read the Tensor hooks here.

Comment (albanD) [inline torch/csrc/autograd/functions/accumulate_grad.h:45]: Shouldn't we have the same for retain_grad hook? Or is it assumed that leaf Tensors can't have retain_grad hooks?

File: torch/csrc/autograd/input_buffer.h [hunk 3]
```diff
@@ -43,7 +43,6 @@ struct InputBuffer {
   // Returns the inputs as a list of variables. Destroys given InputBuffer.
   static std::vector<Variable> variables(InputBuffer&& g);
 
- private:
```
Comment (albanD) [inline torch/csrc/autograd/input_buffer.h:46]: Why?

File: torch/csrc/autograd/python_cpp_function.cpp [hunk 4]
```diff
@@ -77,11 +77,16 @@ PyObject* THPCppFunction_call(
 
 int THPCppFunction_traverse(PyObject* self, visitproc visit, void* arg) {
   auto& fn = *((THPCppFunction*)self)->cdata;
-  for (const auto& hook : fn.pre_hooks()) {
+  for (const auto& hook : fn.tensor_pre_hooks()) {
```
Comment (albanD) [inline torch/csrc/autograd/python_cpp_function.cpp:80]: retain grad hooks?

File: torch/csrc/autograd/python_function.cpp [hunk 5]
```diff
@@ -213,11 +213,16 @@ static int THPFunction_traverse(THPFunction* self, visitproc visit, void* arg) {
   // TODO: I'm not really sure if we're actually obligated to traverse PyObject
   // that is stored in PyNode, since we don't really own that C++ object.
   if (auto cdata = self->cdata.lock()) {
-    for (const auto& hook : cdata->pre_hooks()) {
+    for (const auto& hook : cdata->tensor_pre_hooks()) {
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:216]: retain grad hook?

File: torch/csrc/autograd/python_variable.cpp [hunk 6]
```diff
@@ -655,6 +655,7 @@ static int THPVariable_clear(THPVariable* self) {
       if (auto grad_acc =
               torch::autograd::impl::try_get_grad_accumulator(tensor)) {
         grad_acc->pre_hooks().clear();
+        grad_acc->tensor_pre_hooks().clear();
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:658]: retain grad hook

File: torch/csrc/autograd/python_variable.cpp [hunk 7]
```diff
@@ -2140,10 +2140,12 @@ static int THPVariable_subclass_traverse(
           }
         }
       }
-
-      for (const auto& hook : torch::autograd::impl::hooks(tensor)) {
-        if (auto pyhook = dynamic_cast<PyFunctionTensorPreHook*>(hook.get())) {
-          Py_VISIT(pyhook->dict);
+      if (autograd_meta) {
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:2143]: Why is this needed? The torch::autograd::impl::hooks call used to handle the case without autograd_meta but doesn't anymore?

File: torch/csrc/autograd/variable.cpp [hunk 8]
```diff
@@ -167,27 +167,23 @@ void update_cpp_hooks_on_new_gradfn(
...
-  meta->retains_grad_ = 0;
-  std::unique_ptr<FunctionPreHook> hook_ptr(
-      new CppFunctionPreHook(meta->cpp_hooks_list_, self.output_nr()));
-  new_fn->add_pre_hook(std::move(hook_ptr));
+  if (self.retains_grad()) {
+    auto new_list = std::make_shared<hooks_list>();
+    new_list->push_back(std::move((*meta->retains_grad_hooks_list_)[0]));
+    (*meta->retains_grad_hooks_list_)[0] = nullptr;
+    meta->retains_grad_hooks_list_ = new_list;
+    std::unique_ptr<FunctionPreHook> hook_ptr(new CppFunctionTensorPreHook(
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:183]: Shouldn't you use make_unique instead of new here?

File: torch/csrc/autograd/variable.cpp [hunk 9]
```diff
@@ -167,27 +167,23 @@ void update_cpp_hooks_on_new_gradfn(
...
-  new_list->push_back(std::move((*meta->cpp_hooks_list_)[idx]));
-  (*meta->cpp_hooks_list_)[idx] = nullptr;
-  meta->cpp_hooks_list_ = new_list;
-  // Since this is a new list, 0 is the index of the retains_grad hook
-  meta->retains_grad_ = 0;
-  std::unique_ptr<FunctionPreHook> hook_ptr(
-      new CppFunctionPreHook(meta->cpp_hooks_list_, self.output_nr()));
-  new_fn->add_pre_hook(std::move(hook_ptr));
+  if (self.retains_grad()) {
+    auto new_list = std::make_shared<hooks_list>();
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:179]: This is quite confusing :/
Do you think making the retain_grad hook field not a vector would simplify things?

File: torch/csrc/autograd/variable.cpp [hunk 10]
```diff
@@ -221,17 +217,32 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
...
+void create_cpp_hook(const at::TensorBase& self, bool is_retains_grad_hook) {
   const auto& fn = self.grad_fn();
-  if (fn) {
-    fn->add_pre_hook(std::move(hook_ptr));
+  if (is_retains_grad_hook) {
+    std::shared_ptr<hooks_list>& list =
+        materialize_autograd_meta(self)->retains_grad_hooks_list_;
+    // NOLINTNEXTLINE(modernize-make-shared)
+    list.reset(new hooks_list());
+    std::unique_ptr<FunctionPreHook> hook_ptr{
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:227]: nit: everything above this is shared in both branches, so that can be moved outside of the if statement.

File: torch/csrc/autograd/variable.cpp [hunk 11]
```diff
@@ -342,23 +353,15 @@ const c10::VariableVersion& version_counter(const Variable& self) {
 
 void add_hook(
     const at::TensorBase& self,
-    std::shared_ptr<FunctionPreHook> hook) {
-  materialize_autograd_meta(self)->hooks_.push_back(std::move(hook));
+    std::unique_ptr<FunctionPreHook> hook) {
+  AutogradMeta* meta = materialize_autograd_meta(self);
+  TORCH_INTERNAL_ASSERT(meta->hooks_.size() == 0);
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:358]: Why? You should be able to add multiple hooks no?

File: torch/csrc/autograd/variable.cpp [hunk 12]
```diff
@@ -519,6 +522,24 @@ int64_t VariableHooks::_version(const at::TensorBase& self) const {
   return self.unsafeGetTensorImpl()->version_counter().current_version();
 }
 
+unsigned register_retains_grad_hook(
+    const at::TensorBase& self,
+    std::function<at::TensorBase(const at::TensorBase&)> hook) {
+  TORCH_CHECK(
+      self.requires_grad(),
+      "cannot register a hook on a variable that "
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:530]: cannot retain grad?

File: torch/csrc/autograd/variable.cpp [hunk 13]
```diff
@@ -553,15 +573,16 @@ void VariableHooks::retain_grad(const at::TensorBase& self) const {
         var->mutable_grad() = var->grad() + grad;
       }
     }
+    return at::TensorBase{};
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:576]: What is the contract on the returned value here?

File: torch/csrc/autograd/variable.h [hunk 14]
```diff
@@ -216,16 +217,25 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
...
+  // The hooks_ field is actually reused by both python and cpp logic
+  // For both cases, we have a data structure, cpp_hooks_list_ (cpp)
+  // or dict (python) which is the canonical copy.
+  // Then, for both cases, we always register a single hook to
+  // hooks_ which wraps all the hooks in the list/dict.
+  // And, again in both cases, if the grad_fn exists on that tensor
+  // we will additionally register a single hook to the grad_fn.
+  //
+  // Note that the cpp and python use cases aren't actually aware of
+  // each other, so using both is not defined behavior.
```
Comment (albanD) [inline torch/csrc/autograd/variable.h:229]: What would happen in practice before this PR? They will just run in arbitrary order?
What about after this PR?

File: test/test_autograd.py [hunk 15]
```diff
@@ -845,6 +845,144 @@ def __call__(self, grad):
...
+        a = torch.tensor(1., requires_grad=True)
+        b = a.clone()
+        b.retain_grad()
+        b.register_hook(tensor_prehook)
+
+        b.clone().backward()
+        self.assertEqual(b.grad.item(), 2)
+
+    def test_accumulate_grad_posthooks_can_observe_tensor_prehook(self):
+        # Post hooks on accumulate should be able to observe changes to
```
Comment (albanD) [inline test/test_autograd.py:911]: That should be true of pre hooks on accumulate grad as well right?

File: test/test_autograd.py [hunk 16]
```diff
@@ -845,6 +845,144 @@ def __call__(self, grad):
...
+
+        b = a.clone()
+        acc = b.grad_fn.next_functions[0][0]
+        acc.register_hook(posthook)
+        a.register_hook(tensor_prehook)
+
+        b.backward()
+
+    def test_hook_edge_case_when_called_with_grad(self):
+        # grad executes the tensor prehooks of the next node but not
```
Comment (albanD) [inline test/test_autograd.py:930]: nit: tensor prehooks -> tensor hooks
there is no concept of pre/post hooks for Tensors.

**Review by albanD — COMMENTED** [review]
Sounds good.
This is BC-breaking though?

File: torch/csrc/autograd/variable.cpp (hunk 9 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.cpp:179]: Yes, I was planning on doing that in a follow up, but I can also update that in this PR.

File: torch/csrc/autograd/variable.cpp (hunk 11 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.cpp:358]: Yes, but there shouldn't be a reason to. You shouldn't really be working with the hooks_ field directly. You should either be registering cpp tensor hooks or python tensor hooks, and those mechanisms should guarantee that there is only one hook in hook_ at a time.

File: torch/csrc/autograd/variable.cpp (hunk 13 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.cpp:576]: If the return value is defined, then it will modify the gradient; otherwise if it is not-defined, then this hook is considered as only "reading" the value of the gradient.

See: https://github.com/pytorch/pytorch/blob/645fb217c06348a4f1ccdf68a93bd711f7158c62/torch/csrc/autograd/cpp_hook.cpp#L38-L40

File: torch/csrc/autograd/functions/accumulate_grad.h (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/functions/accumulate_grad.h:45]: Yes, I think it is ok to assume that.

File: torch/csrc/autograd/input_buffer.h (hunk 3 above)
Comment (soulitzer) [reply torch/csrc/autograd/input_buffer.h:46]: Let me see if I can avoid this.

File: torch/csrc/autograd/python_variable.cpp (hunk 7 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_variable.cpp:2143]: There's no reason to call hooks if there is no autograd_meta. I felt like it would be clearer if we enforce that when we call it. I don't feel strongly about this though.

File: torch/csrc/autograd/variable.cpp (hunk 10 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.cpp:227]: Yeah this one looks kinda ugly right now with so much repetition, but I'm not sure how to simplify further at the moment - the top branch uses `retains_grad_hooks_list_`, and the second has `cpp_hooks_list_`. It's tricky because references must be initialized when defined and cannot be mutated to point to different things.

File: torch/csrc/autograd/variable.cpp (hunk 9 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.cpp:179]: Since this is the way it was before, are you okay with keeping this for now, and leaving that as a follow up. If not, let me know.

File: torch/csrc/autograd/variable.h (hunk 14 above)
Comment (soulitzer) [reply torch/csrc/autograd/variable.h:229]: From the python side the hooks_ field is cleared everytime backward_hooks field is set, and a PyFunctionHook is registered.
From the cpp side, hooks_ field is set whenever a cpp hook is registered AND cpp_hooks_list_ is empty. (In the case where we register a retains_grad hook that is also the case, but it shouldn't matter because hooks_ is only used on leaf nodes and retains_grad is no-op when invoked on leaf node)

What happens is that one will be overwritten by the other.

This is still the case after this PR, but I'm okay with not caring about mixed Python + cpp case.

File: test/test_autograd.py (hunk 16 above)
Comment (soulitzer) [reply test/test_autograd.py:930]: Will need another follow up to purge all the existing usages of tensor "pre"hook

File: torch/csrc/autograd/variable.cpp (hunk 9 above)
Comment (albanD) [reply torch/csrc/autograd/variable.cpp:179]: Sure, could you open a BE issue so that we don't forget?

File: torch/csrc/autograd/variable.h (hunk 14 above)
Comment (albanD) [reply torch/csrc/autograd/variable.h:229]: Ho wow this is very badly broken :O good to know!

**Review by albanD — APPROVED** [review]
lint needs fixing but sounds good otherwise!

File: torch/csrc/autograd/input_buffer.h (hunk 3 above)
Comment (soulitzer) [reply torch/csrc/autograd/input_buffer.h:46]: I don't think we can easily avoid this for now. It should be gone for sure though when we eventually remove the usage of captures hooks, so should just be a temporary hack.

File: torch/csrc/autograd/function.h [hunk 17]
```diff
@@ -506,6 +514,15 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     return pre_hooks_;
   }
 
+  virtual std::vector<std::unique_ptr<FunctionPreHook>>&
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/function.h:517]: there's a class that inherits from Node that overrides tensor_pre_hooks_
